File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/api/task/CommunicationGroupClient.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.reef.io.network.group.api.task;

import org.apache.reef.annotations.audience.TaskSide;
import org.apache.reef.io.network.group.api.operators.Broadcast;
import org.apache.reef.io.network.group.api.operators.Gather;
import org.apache.reef.io.network.group.api.operators.Reduce;
import org.apache.reef.io.network.group.api.GroupChanges;
import org.apache.reef.io.network.group.api.operators.Scatter;
import org.apache.reef.io.network.group.impl.driver.TopologySimpleNode;
import org.apache.reef.io.network.group.impl.task.CommunicationGroupClientImpl;
import org.apache.reef.tang.annotations.DefaultImplementation;
import org.apache.reef.tang.annotations.Name;
import org.apache.reef.wake.Identifier;

import java.util.List;

/**
 * The Task side interface of a communication group.
 * Lets one get the operators configured for this task
 * and use them for communication between tasks configured
 * in this communication group
 */
@TaskSide
@DefaultImplementation(value = CommunicationGroupClientImpl.class)
public interface CommunicationGroupClient {

  /**
   * @return The name configured on this communication group
   */
  Class<? extends Name<String>> getName();

  /**
   * The broadcast sender configured on this communication group.
   * with the given oepratorName
   *
   * @param operatorName
   * @return
   */
  Broadcast.Sender getBroadcastSender(Class<? extends Name<String>> operatorName);

  /**
   * The broadcast receiver configured on this communication group.
   * with the given oepratorName
   *
   * @param operatorName
   * @return
   */
  Broadcast.Receiver getBroadcastReceiver(Class<? extends Name<String>> operatorName);

  /**
   * The reduce receiver configured on this communication group.
   * with the given operatorName
   *
   * @param operatorName
   * @return
   */
  Reduce.Receiver getReduceReceiver(Class<? extends Name<String>> operatorName);

  /**
   * The reduce sender configured on this communication group.
   * with the given operatorName
   *
   * @param operatorName
   * @return
   */
  Reduce.Sender getReduceSender(Class<? extends Name<String>> operatorName);

  /**
   * Return the scatter sender configured on this communication group.
   * {@code operatorName} is used to specify the scatter sender to return.
   *
   * @param operatorName
   * @return
   */
  Scatter.Sender getScatterSender(Class<? extends Name<String>> operatorName);

  /**
   * Return the scatter receiver configured on this communication group.
   * {@code operatorName} is used to specify the scatter receiver to return.
   *
   * @param operatorName
   * @return
   */
  Scatter.Receiver getScatterReceiver(Class<? extends Name<String>> operatorName);

  /**
   * Return the gather receiver configured on this communication group.
   * {@code operatorName} is used to specify the gather receiver to return.
   *
   * @param operatorName
   * @return
   */
  Gather.Receiver getGatherReceiver(Class<? extends Name<String>> operatorName);

  /**
   * Return the gather sender configured on this communication group.
   * {@code operatorName} is used to specify the gather sender to return.
   *
   * @param operatorName
   * @return
   */
  Gather.Sender getGatherSender(Class<? extends Name<String>> operatorName);

  /**
   * @return Changes in topology of this communication group since the last time
   * this method was called
   */
  GroupChanges getTopologyChanges();

  /**
   * @return list of current active tasks, last updated during updateTopology()
   */
  List<Identifier> getActiveSlaveTasks();

  /**
   * @return root node of simplified topology representation
   */
  TopologySimpleNode getTopologySimpleNodeRoot();

  /**
   * Asks the driver to update the topology of this communication group. This can
   * be an expensive call depending on what the minimum number of tasks is for this
   * group to function as this first tells the driver, driver then tells the affected
   * tasks and the driver gives a green only after affected tasks have had a chance
   * to be sure that their topology will be updated before the next message is
   * communicated
   */
  void updateTopology();


}

File: lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/CommunicationGroupClient.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System;
using System.Collections.Generic;
using System.Reflection;
using Org.Apache.REEF.Network.Group.Config;
using Org.Apache.REEF.Network.Group.Operators;
using Org.Apache.REEF.Network.Group.Operators.Impl;
using Org.Apache.REEF.Tang.Annotations;
using Org.Apache.REEF.Tang.Formats;
using Org.Apache.REEF.Tang.Interface;
using Org.Apache.REEF.Tang.Util;
using Org.Apache.REEF.Utilities.Diagnostics;
using Org.Apache.REEF.Utilities.Logging;

namespace Org.Apache.REEF.Network.Group.Task.Impl
{
    /// <summary>
    ///  Used by Tasks to fetch Group Communication Operators in the group configured by the driver.
    /// </summary>
    internal sealed class CommunicationGroupClient : ICommunicationGroupClientInternal
    {
        private static readonly Logger LOGGER = Logger.GetLogger(typeof(CommunicationGroupClient));
        private readonly Dictionary<string, object> _operators;

        /// <summary>
        /// Creates a new CommunicationGroupClient.
        /// </summary>
        /// <param name="groupName">The name of the CommunicationGroup</param>
        /// <param name="operatorConfigs">The serialized operator configurations</param>
        /// <param name="groupCommNetworkObserver">The handler for all incoming messages
        /// across all Communication Groups</param>
        /// <param name="configSerializer">Used to deserialize operator configuration.</param>
        /// <param name="commGroupNetworkHandler">Communication group network observer that holds all the handlers for each operator.</param>
        /// <param name="injector">injector forked from the injector that creates this instance</param>
        [Inject]
        private CommunicationGroupClient(
            [Parameter(typeof(GroupCommConfigurationOptions.CommunicationGroupName))] string groupName,
            [Parameter(typeof(GroupCommConfigurationOptions.SerializedOperatorConfigs))] ISet<string> operatorConfigs,
            IGroupCommNetworkObserver groupCommNetworkObserver,
            AvroConfigurationSerializer configSerializer,
            ICommunicationGroupNetworkObserver commGroupNetworkHandler,
            IInjector injector)
        {
            _operators = new Dictionary<string, object>();

            GroupName = groupName;
            groupCommNetworkObserver.Register(groupName, commGroupNetworkHandler);

            foreach (string operatorConfigStr in operatorConfigs)
            {                
                IConfiguration operatorConfig = configSerializer.FromString(operatorConfigStr);

                IInjector operatorInjector = injector.ForkInjector(operatorConfig);
                string operatorName = operatorInjector.GetNamedInstance<GroupCommConfigurationOptions.OperatorName, string>(
                    GenericType<GroupCommConfigurationOptions.OperatorName>.Class);
                string msgType = operatorInjector.GetNamedInstance<GroupCommConfigurationOptions.MessageType, string>(
                    GenericType<GroupCommConfigurationOptions.MessageType>.Class);

                Type groupCommOperatorGenericInterface = typeof(IGroupCommOperator<>);
                Type groupCommOperatorInterface = groupCommOperatorGenericInterface.MakeGenericType(Type.GetType(msgType));
                var operatorObj = operatorInjector.GetInstance(groupCommOperatorInterface);
                _operators.Add(operatorName, operatorObj);
            }
        }

        /// <summary>
        /// Returns the Communication Group name
        /// </summary>
        public string GroupName { get; private set; }

        /// <summary>
        /// Gets the BroadcastSender with the given name and message type.
        /// </summary>
        /// <typeparam name="T">The message type</typeparam>
        /// <param name="operatorName">The name of the Broadcast operator</param>
        /// <returns>The BroadcastSender</returns>
        public IBroadcastSender<T> GetBroadcastSender<T>(string operatorName)
        {
            return GetOperatorInstance<BroadcastSender<T>>(operatorName);
        }

        /// <summary>
        /// Gets the BroadcastReceiver with the given name and message type.
        /// </summary>
        /// <typeparam name="T">The message type</typeparam>
        /// <param name="operatorName">The name of the Broadcast operator</param>
        /// <returns>The BroadcastReceiver</returns>
        public IBroadcastReceiver<T> GetBroadcastReceiver<T>(string operatorName)
        {
            return GetOperatorInstance<BroadcastReceiver<T>>(operatorName);
        }

        /// <summary>
        /// Gets the ReduceSender with the given name and message type.
        /// </summary>
        /// <typeparam name="T">The message type</typeparam>
        /// <param name="operatorName">The name of the Reduce operator</param>
        /// <returns>The ReduceSender</returns>
        public IReduceSender<T> GetReduceSender<T>(string operatorName)
        {
            return GetOperatorInstance<ReduceSender<T>>(operatorName);
        }

        /// <summary>
        /// Gets the ReduceReceiver with the given name and message type.
        /// </summary>
        /// <typeparam name="T">The message type</typeparam>
        /// <param name="operatorName">The name of the Reduce operator</param>
        /// <returns>The ReduceReceiver</returns>
        public IReduceReceiver<T> GetReduceReceiver<T>(string operatorName)
        {
            return GetOperatorInstance<ReduceReceiver<T>>(operatorName);
        }

        /// <summary>
        /// Gets the ScatterSender with the given name and message type.
        /// </summary>
        /// <typeparam name="T">The message type</typeparam>
        /// <param name="operatorName">The name of the Scatter operator</param>
        /// <returns>The ScatterSender</returns>
        public IScatterSender<T> GetScatterSender<T>(string operatorName)
        {
            return GetOperatorInstance<ScatterSender<T>>(operatorName);
        }

        /// <summary>
        /// Gets the ScatterReceiver with the given name and message type.
        /// </summary>
        /// <typeparam name="T">The message type</typeparam>
        /// <param name="operatorName">The name of the Scatter operator</param>
        /// <returns>The ScatterReceiver</returns>
        public IScatterReceiver<T> GetScatterReceiver<T>(string operatorName)
        {
            return GetOperatorInstance<ScatterReceiver<T>>(operatorName);
        }

        /// <summary>
        /// Gets the Group Communication operator with the specified name and type.
        /// If the operator hasn't been instanciated yet, find the injector 
        /// associated with the given operator name and use the type information 
        /// to create a new operator of that type.
        /// </summary>
        /// <typeparam name="T">The type of operator to create</typeparam>
        /// <param name="operatorName">The name of the operator</param>
        /// <returns>The newly created Group Communication Operator</returns>
        private T GetOperatorInstance<T>(string operatorName) where T : class
        {
            if (string.IsNullOrEmpty(operatorName))
            {
                throw new ArgumentNullException("operatorName");
            }

            object op;
            if (!_operators.TryGetValue(operatorName, out op))
            {
                Exceptions.Throw(new ArgumentException("Operator is not added at Driver side:" + operatorName), LOGGER);
            }

            return (T)op;
        }

        /// <summary>
        /// Call each Operator to ensure all the nodes in the topology group has been registered
        /// </summary>
        void ICommunicationGroupClientInternal.WaitingForRegistration()
        {
            foreach (var op in _operators.Values)
            {
                var method = op.GetType().GetMethod("Org.Apache.REEF.Network.Group.Operators.IGroupCommOperatorInternal.WaitForRegistration", BindingFlags.NonPublic | BindingFlags.Instance);
                method.Invoke(op, null);
            }
        }
    }
}
